presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java

```diff
@@ -61,13 +62,17 @@
 
     private final SqlStageExecution stage;
     private final List<InternalNode> nodes;
-    private final List<SourceScheduler> sourceSchedulers;
+
+    // use a CopyOnWriteArrayList to prevent ConcurrentModificationExceptions
+    // if close() is called while the main thread is in the scheduling loop
+    private final CopyOnWriteArrayList<SourceScheduler> sourceSchedulers;
     private final List<ConnectorPartitionHandle> partitionHandles;
     private boolean scheduledTasks;
     private boolean anySourceSchedulingFinished;
     private final Optional<LifespanScheduler> groupedLifespanScheduler;
 
     private final Queue<Integer> tasksToRecover = new ConcurrentLinkedQueue<>();
+    private volatile boolean cancelled;
```

Comment:
I don't think you need volatile here, but you should annotate it with `@GuardedBy("this")`